distributed/shuffle/tests/test_shuffle.py

```diff
@@ -2724,3 +2698,56 @@ async def test_barrier_handles_stale_resumed_transfer(c, s, *workers):
     await wait_for_tasks_in_state("shuffle-transfer", "resumed", 1, barrier_worker)
     barrier_worker.block_gather_dep.set()
     await out
+
+
+@pytest.mark.parametrize("disk", [True, False])
+@pytest.mark.parametrize("keep", ["first", "last"])
+@gen_cluster(client=True)
+async def test_shuffle_stable_ordering(c, s, a, b, keep, disk):
+    """Ensures that shuffling guarantees ordering for individual entries
+    belonging to the same shuffle key"""
+    df = dask.datasets.timeseries(
+        start="2000-01-01",
+        end="2000-02-01",
+        dtypes={"x": int, "y": int, "z": int},
+        freq="1 s",
+    )
+    df["x"] = df["x"] % 23
+    df["y"] = df["y"] % 19
+    df["z"] = df["z"] % 17
```

Comment:
Maybe it is easier to construct a dataframe very explicitly and assert explicitly on the output group instead of testing this indirectly via drop_duplicates.

Reply:
I'm not sure if it's easier to write the test, but I see your point that it might be easier to argue about it.